nilearn/regions/signal_extraction.py

```diff
@@ -15,11 +15,162 @@
 from .. import masking
 from ..image import new_img_like
 
+INF = 1000 * np.finfo(np.float32).eps
+
+
+def _check_shape_affine_label_img(target_img, labels_img):
+    """Validate shapes and affines of labels.
+
+    Parameters
+    ----------
+    target_img : Niimg-like object
+        See :ref:`extracting_data`.
+        Image to extract the data from.
+
+    labels_img : Niimg-like object
+        See :ref:`extracting_data`.
+        Regions definition as labels.
+        Encodes the region labels of the signals.
+
+    """
+    target_affine = target_img.affine
+    target_shape = target_img.shape[:3]
+
+    def err_msg(x):
+        return f"labels_img and target_img must have same {x}."
+
+    if labels_img.shape != target_shape:
+        raise ValueError(err_msg("shape"))
+    if (labels_img.affine.shape != target_affine.shape
+            or abs(labels_img.affine - target_affine).max() > INF):
+        raise ValueError(err_msg("affine"))
+
+
```

Comment:
I might missing something, but I'm wondering whether we could combine `_check_shape_affine_label_img` and `_check_shape_affine_maps_masks` in a more generic function.
Also, it is not super clear to me why the `dim` parameter is useful. Isn't it always set to 3 ?

Maybe something along those lines:

```python
def _check_shape_and_affine_compatibility(img1, img2, dim=3):
    _check_shape_compatibility(img1, img2, dim=dim)
    _check_affine_equality(img1, img2)

def _check_shape_compatibility(img1, img2, dim=3):
     """Check that shapes match for dimensions going from 0 to dim-1."""
     if img1.shape[:dim] != img2.shape[:dim]:
         raise ValueError(f"{img1} and {img2} have incompatible shapes.")

def _check_affine_equality(img1, img2):
     if (
         img1.affine.shape != img2.affine.shape
         or abs(img1.affine - img2.affine).max() > INF
     ):
         raise ValueError(f"{img1} and {img2} have different affine matrices.")
```

And in `_get_labels_data`:

```python
_check_shape_and_affine_compatibility(target_img, labels_img)
....
if mask_img is not None:
    _check_shape_and_affine_compatibility(target_img, mask_img)
    mask_img = _utils.check_niimg_3d(mask_img)
    mask_data = _safe_get_data(mask_img, ensure_finite=True)
    ...
```

Reply:
Tried to do a bit of refactoring as you suggested.